src/distributions/slice.rs

```diff
@@ -81,6 +83,13 @@ impl<'a, T> Slice<'a, T> {
             }),
         }
     }
+
+    /// Returns the count of choices in this distribution
+    pub fn choices(&self) -> NonZeroUsize {
+        // Safety: at construction time, it was ensured that the slice was
+        // non-empty, as such the length can never be 0.
+        unsafe { NonZeroUsize::new_unchecked(self.slice.len()) }
```

Comment:
Is there a strong rationale for inclusion of `unsafe` here? Why not just return `usize`?

Reply:
I mean as part of the struct creation we have the contract that the length can never be 0, as such I'd expect the choices() type to represent that as well. (Which allows for some optimizations from the rust compiler) As an alternative I would probably use `NonZeroUsize::new(self.slice.len()).unwrap()` but I think that isn't really nicer with the `.unwrap()` in there.

The distribution implementation also already makes the same trade-off (not `.unwrap()`-ing), as such I would argue it's valid to do the same thing here as well.

Reply:
well as I already argued above having it NonZeroUsize makes it abundantly clear in the function signature to the caller that it is never zero, which allows for for example an unprotected divide ( 1 / self.num_choices() ) without worrying that that will ever return infinity ( and as such break stuff ) , which is also relevant here since that represents the probability of every element. 

another point may be if you have something like Option<Num_Choices> Option<NonZeroUsize> is sizeof(usize) and Option<usize> is sizeof(usize) + 1 byte

Reply:
I personally would love to have a NonZeroUsize here, but I think in the end it comes down a bit to personal preference I guess. I mean the unsafe here is also really easy to reason about so I wouldn't worry about bugs introduced by it.

Reply:
I would prefer the unwrap here, IMHO using unsafe code here would have to be justified by observable performance problems.

Reply:
https://github.com/rust-random/rand/pull/1402/commits/3094c50c3bae772bbd502b7bdd4290a4bce9b211 how about this? this avoids both unsafe and unwrap at the cost of an additional struct field.